### ISSUE TITLE
should we provide uucore as separated cargo library?

### ISSUE THREAD

**Issue by mpkh** [uutils/coreutils#931]
and use it inside coreutils like git submodule or cargo package with fixed version?

why I think it could be useful:

basically if someone will like to use same functional as provided by some coreutil he will currently be forced to overtake either code from util and code from uucore, having it as cargo library it will be possible for him to avoid second copy-paste and getting for us potential contributor of this library.

**Comment by wimh** [comment]
I am new to rust, so you should probably not listen to me. However I can give some comments....

* If uucore is moved into a different git repo, you will probably lose some history (git blame). 
* uucore seems to contain miscellaneous code. I am not sure if someone would find uucore on crates.io if the description does not match the content. For example would I find it if search for something to parse the time? I would probably look into chrono or humantime instead which are search results 1 and 3.
* If someone finds something useful, he/she might just ask for a solution at that moment. At least that's what I would [do](https://github.com/dtolnay/indoc/issues/2#issuecomment-233203263).
* Besides uucore, other parts might also be useful. For example if someone writes an online text editor, and needs a wordcount, the code from wc might be useful, but not as an executable, but an api which runs in-memory. 
* I personally would prefer to move code into separate, dedicated crates, for example that what I am currently trying with [mockstream.rs](https://github.com/wimh/coreutils/blob/od/src/od/mockstream.rs#L1).
* But that does not always work out; so [byteorder_io.rs](https://github.com/wimh/coreutils/blob/od/src/od/byteorder_io.rs#L1) contains copied code (slightly modified)....

**Comment by frewsxcv** [comment]
> If uucore is moved into a different git repo, you will probably lose some history (git blame).

For the record, [it is possible to retain git history for this scenario](https://help.github.com/articles/splitting-a-subfolder-out-into-a-new-repository/).

**Comment by jbcrail** [comment]
My original intent for creating `uucore` was to eventually have a separate Cargo package with a separate development timeline.

Pros include (as noted by @mpkh):
* improved testing (none at the moment)
* central point to encapsulate cross-platform issues (ideally utilities could be platform agnostic)
* library for [common options](http://www.gnu.org/software/coreutils/manual/html_node/Common-options.html#Common-options): block sizes, users, groups, etc.
* move dependencies into `uucore`, simplifying the `uutils` build

Cons include:
* harder for new contributors to create utilities (ameliorated with help from core developers)

Given that Servo and other projects have separate crates for their own dependencies, I'm not too concerned about confusion about the package's usage. If we advertise `uucore` simply as a library for creating cross-platform command-line applications with `uutils` as the primary user, then we may gain interested users.

@frewsxcv Thanks for the info about retaining git history!

**Comment by knight42** [comment]
Providing `uucore` as a separated library sounds good to me, but personally I would postpone it, because `uucore` has not been stable:

* The declarations of C structs and functions in `uucore`, especially in `uucore::c_types` and `uucore::utmpx` should be moved to `libc`
* New modules may be added to `uucore` in the future as the whole project hasn't finished

Therefore separating `uucore` leaves it a bit inconvenient to make the changes and I would consider to separate it only after we finish the whole project.

**Comment by nathanross** [comment]
I have two opinions and a question.

#### opinion 1. Spinning out uucore into a separate repository is fine with me, I do not think this decision needs to wait on questions of whether and what audience uucore will have outside of uutils.

#### opinion 2. Deciding to change / not change uucore's mission/audience can be considered a separate decision, and I think it is one where we benefit from first deciding what we're trying to achieve with uucore (and, in a broader sense, the uutils project).

         # Are we trying to create the "best possible" coreutils implementation (for a given compromise of memory efficiency, CPU efficiency, and code maintainability - within which I'm including "simplicity") - without *any* regard to how much it would help or advance other rust-CLI-utilities.

         # or to create a really really good coreutils, but also spend project time, code complexity, code CPU and allocation efficiency on scenarios that don't necessarily help coreutils, but make direct strides towards advancing a larger goal of a rust-CLI-utility userland

      I have no answer or opinion on this question. I think both goals would be really really cool but I acknowledge they are mutually exclusive. 

      Knowing our goal would allow us to advance towards it faster, by moving towards it consistently across all decisions where these two goals diverge (uucore library design, level of focus on tooling reusability, organizational effort made or avoided on outreach and invitation of other nascent rust projects reimplementing CLI-utilities) .

**Comment by mpkh** [comment]
from current positions I think it's better not to change at the moment. @nathanross yes I think it could  be another question/issue/poll...
